v3/lints/cabf_br/lint_sub_cert_eku_check.go

```diff
@@ -0,0 +1,83 @@
+package cabf_br
+
+/*
+ * ZLint Copyright 2024 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import (
+	"fmt"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v3/lint"
+	"github.com/zmap/zlint/v3/util"
+)
+
+type subExtKeyUsageCheck struct{}
+
+func init() {
+	lint.RegisterCertificateLint(&lint.CertificateLint{
+		LintMetadata: lint.LintMetadata{
+			Name:          "e_sub_cert_eku_check",
+			Description:   "Subscriber certificates MUST have id-kp-serverAuth and MAY have id-kp-clientAuth present in extKeyUsage",
+			Citation:      "BRs: 7.1.2.7.10 Subscriber Certificate Extended Key Usage",
+			Source:        lint.CABFBaselineRequirements,
+			EffectiveDate: util.CABFBRs_2_0_0_Date,
+		},
+		Lint: NewSubExtKeyUsageCheck,
+	})
+}
+
+func NewSubExtKeyUsageCheck() lint.LintInterface {
+	return &subExtKeyUsageCheck{}
+}
+
+func (l *subExtKeyUsageCheck) CheckApplies(c *x509.Certificate) bool {
+	return util.IsSubscriberCert(c) && util.IsExtInCert(c, util.EkuSynOid)
+}
+
+func (l *subExtKeyUsageCheck) Execute(c *x509.Certificate) *lint.LintResult {
+	var hasClientAuthEKU, hasServerAuthEKU bool
+
+	for _, eku := range c.ExtKeyUsage {
+		switch eku {
+		case x509.ExtKeyUsageServerAuth:
+			hasServerAuthEKU = true
+
+		case x509.ExtKeyUsageClientAuth:
+			hasClientAuthEKU = true
+
+		case x509.ExtKeyUsageAny, x509.ExtKeyUsageCodeSigning, x509.ExtKeyUsageTimeStamping,
+			x509.ExtKeyUsageOcspSigning, x509.ExtKeyUsageEmailProtection:
+
+			return &lint.LintResult{Status: lint.Error, Details: fmt.Sprintf("%s MUST NOT be present", util.GetEKUString(eku))}
+		}
+	}
+
+	if !hasServerAuthEKU {
+		return &lint.LintResult{Status: lint.Error, Details: "id-kp-serverAuth MUST be present"}
+	}
+
+	if len(c.UnknownExtKeyUsage) > 0 {
+		for _, eku := range c.UnknownExtKeyUsage {
+			if eku.Equal(util.PreCertificateSigningCertificateEKU) {
+				return &lint.LintResult{Status: lint.Error, Details: "Precertificate Signing Certificate extKeyUsage MUST NOT be present"}
+			}
+		}
+	}
+
+	if (len(c.ExtKeyUsage) > 2 && !hasClientAuthEKU) || len(c.UnknownExtKeyUsage) > 0 {
```

Comment:
Maybe you could set a variable in the `default` case in the `range c.ExtKeyUsage` loop and check that.

Reply:
I may advocate for keeping a list of of OIDs found during the execution of the lint and cross referencing them at them end against the list of known OIDs.

```go
var known = map[string]struct{}{
	"1.3.6.1.5.5.7.3.1": {},
	"1.3.6.1.5.5.7.3.2": {},
	"1.3.6.1.5.5.7.3.3": {},
	"1.3.6.1.5.5.7.3.4": {},
	"1.3.6.1.5.5.7.3.8": {},
	"1.3.6.1.5.5.7.3.9": {},
	"2.5.29.37.0": {},
	"1.3.6.1.4.1.11129.2.4.4": {},
}

func lint() {
    found := []string
    for _, eku := range c.ExtKeyUsage {
        found = append(found, eku.String()
        ....
        ....
        ....
    }
    for _, eku := range.UnknownExtKeyUsage {
        found = append(found, eku.String()
        ....
        ....
        ....
    }
    for _, oid := range found {
        _, ok := known[oid]
        if !ok {
            return &lint.LintResult{Status: lint.Warn, Details: "any other value than id-kp-serverAuth and id-kp-clientAuth is NOT RECOMMENDED"}
        }
    }
}

```

Reply:
The ExtKeyUsage and UnknownExtKeyUsage are not of the same type:

```go
ExtKeyUsage        []ExtKeyUsage           // Sequence of extended key usages.
UnknownExtKeyUsage []asn1.ObjectIdentifier // Encountered extended key usages unknown to this package.
```

The lint currently checks for MUST, MUST NOT, MAY, and NOT RECOMMENDED, and returns the corresponding lint status and error.

We can create a map with each EKU and the requirement (i.e., MUST) and lookup the value while ranging over the two EKU lists, but I guess this would be less efficient

Reply:
I reckon that this would be an efficiency modifier in the range of nanoseconds (possibly into ms) for each run of the lint. In general, legibility trumps such optimizations since correctness of the lint is paramount.